### ISSUE TITLE
What settings need to be modified to train on my dataset?

### ISSUE THREAD

**Issue by hustCYQ** [open-mmlab/mmdetection3d#1518]
I have created a Kitti format dataset by myself. And I choose pointpillars as a model. Do I only need to modify the following settings?

point_cloud_range = [0, -39.68, -3, 69.12, 39.68, 1] --> point_cloud_range = [-34.56, -39.68, -1, 34.56, 39.68, 3]
bbox_head.anchor_generator.ranges

I modified these settings, but my AP is always 0.

**Comment by hustCYQ** [comment]
And I use test.py --show to visual prediction results and found that only a quarter of the point cloud is displayed.

**Comment by ZCMax** [comment]
First, I suggest you to use the `tools/browsw_dataset.py`  script to visualize your training datasets and labels to check whether your customized dataset is prepared well. Then you can adjust your training config following this tutorial https://github.com/open-mmlab/mmdetection3d/blob/master/docs/en/tutorials/pure_point_cloud_dataset.md#training-setting-adjustment.

**Comment by hustCYQ** [comment]
> First, I suggest you to use the `tools/browse_dataset.py` script to visualize your training datasets and labels to check whether your customized dataset is prepared well. Then you can adjust your training config following this tutorial https://github.com/open-mmlab/mmdetection3d/blob/master/docs/en/tutorials/pure_point_cloud_dataset.md#training-setting-adjustment.

Thank you for your valuable suggestions! I have modified the dataset, and I run tools/browse_dataset.py to ensure that there is no problem with the data. However, I still found that the model handled myself data poorly (ap=0), even though I had checked all the config files. I wonder if there is any hard coding elsewhere?

**Comment by hustCYQ** [comment]
> First, I suggest you to use the `tools/browse_dataset.py` script to visualize your training datasets and labels to check whether your customized dataset is prepared well. Then you can adjust your training config following this tutorial https://github.com/open-mmlab/mmdetection3d/blob/master/docs/en/tutorials/pure_point_cloud_dataset.md#training-setting-adjustment.

And I found that  X direction in point_cloud_range cannot be modified, otherwise the training effect will be very poor. My data range is [-34.56, -39.68, -1, 34.56, 39.68, 3]. Do I miss any settings?

**Comment by ZCMax** [comment]
Have ever visualized the detection results of your trained model? And can you provide the training log to see whether your loss converges well?

**Comment by hustCYQ** [comment]
> Have you ever visualized the detection results of your trained model? And can you provide the training log to see whether your loss converges well?
I visualized the results, but I found that with the training of the model, the prediction bbox did not change. I think the model has not been trained at all. Loss didn't reduce since epoch1. 
Since all points in the Kitti dataset are x>0, I wonder if there are some special settings that make the X direction of my data not negative?

**Comment by hustCYQ** [comment]
> Have you ever visualized the detection results of your trained model? And can you provide the training log to see whether your loss converges well?

I visualized the results, but I found that with the training of the model, the prediction bbox did not change. I think the model has not been trained at all. Loss didn't reduce since epoch1.
Since all points in the Kitti dataset are x>0, I wonder if there are some special settings that make the X direction of my data not negative?

**Comment by hustCYQ** [comment]
> Have you ever visualized the detection results of your trained model? And can you provide the training log to see whether your loss converges well?

![image](https://user-images.githubusercontent.com/39451786/171138841-c4e13acf-0f63-4a94-b526-04929696da1b.png)

![image](https://user-images.githubusercontent.com/39451786/171138923-556ad768-3783-4c01-b9d2-0c5eac67433e.png)

 There are results of running tools/browse_dataset.py.

**Comment by hustCYQ** [comment]
> Have you ever visualized the detection results of your trained model? And can you provide the training log to see whether your loss converges well?

I modify point_cloud_range,  model.voxel_layer.point_cloud_range, model.voxel_encoder.point_cloud_range, bbox_head.anchor_generator.ranges in hv_pointpillars_secfpn_6x8_160e_kitti-3d-car.py. 
Do I miss any settings?

**Comment by ste-wit** [comment]
Did you find out what the cause was? I experience a similar issue.

**Comment by damonftl** [comment]
Count me in as having the same problem too. I always get AP = 0.0 with the objects in my custom dataset that is only point clouds. I have modified the point_cloud_range to be [-20.0, -5.0, 0.0, 20.0, 80.0, 10.0] to match my dataset. I'm using a PointPillars model.

The browse_dataset script shows my point clouds perfectly, together with the ground truth bounding boxes in the right places and around the correct objects.

Loss appears to be doing what it should according to the graphs from tensorboard:

![pointpillars-36epochs-output-9-23-22 (zero accuracy)](https://user-images.githubusercontent.com/61205651/192888140-f98b8d7d-525a-4fe9-a689-c697d02be77e.png)

When running tests using the trained model none of the objects are detected successfully, but it's really hard to tell what's actually going on. AP never gets to anything above 0.0, and the point data and annotations are loaded correctly from what I can tell. I don't quite know what else to double-check.

**Comment by ste-wit** [comment]
In my case (see #1862) the increased z-range of the point_cloud_range resulted in more voxels and hence due to the voxel_limit important voxels were dropped.

This was not visible when using `browse_dataset.py` since here only the data pipeline is processed but no model operations take place. In PointPillars this should not be a problem though. But to validate, put a breakpoint in the forward pass in the model https://github.com/open-mmlab/mmdetection3d/blob/9556958fe1c6fe432d55a9f98781b8fdd90f4e9c/mmdet3d/models/detectors/voxelnet.py#L43 and plot the voxelized point cloud using Open3D:
```python
import open3d as o3d
pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(voxels) # who have to check the shape of voxels and slice/reshape it to be Nx3
o3d.visualization.draw_geometries([pcd])
```

**Comment by damonftl** [comment]
Thank you for the reply! I added this validation as you suggested and I am getting an interesting visualization. Unfortunately I'm not sure if what I'm seeing shows that I have okay voxelization settings or if it shows I need to change something. The voxels do make a shape that looks just like my input data, although there are horizontal gaps (you can see them at bottom left of the image):

![voxel-vis](https://user-images.githubusercontent.com/61205651/193037961-88df6204-fefb-48d3-a4f0-f2bb5f17c7db.png)

I did look at your other issue before and made sure I have the PointShuffle step in my train pipeline. My max_voxels are currently set to (32000, 64000) and voxel_size is [0.5, 0.5, 1.0]. If I go smaller than that voxel size I get index out of bounds errors.

**Comment by damonftl** [comment]
Well, I believe I've found out why I was getting zero AP calculations - in my training data I was setting the "gt_boxes_upright_depth" to just an array of zeroes, since this data is 100% custom and synthetically generated, and depth data is not a part of it. By putting the bounding box position and size (in fact, exactly the same data that goes into "gt_bboxes_3d") in that array I'm now able to get nonzero APs. I'm not even sure that's the right way to go about it with the data I have.

The accuracy it's giving me is absolutely terrible, but I think that's unrelated to this problem. I'm still not sure about the above though, whether the voxels are looking correct or if any adjustments need to be made there.

**Comment by ste-wit** [comment]
I am also not sure whether this is correct but I am also setting `gt_boxes_upright_depth` to the same values as `gt_bboxes_3d`. 

Does anybody else know, whether this is fine or whether this could be the cause of a problem during training?
